src/services/db.js

```diff
@@ -9,136 +10,153 @@ export default class Database {
-          header.map((v,i) => ({
-            [v]: trick[i]
-          }))
-        );
-        tricks[i] = rightFormatTrick;
-      }
+        Object.assign.apply({},
+          header.map((v,i) => {
+            // add 0 for stickFrequency
+            trick.concat([0]);
+            return ({[v]: trick[i]})
+          })
+      ));
 
       // adds the tricks to the database
-      this.db.mainTricks.bulkPut(tricks).then(() =>
-        console.log("added tricks to database from the csv")
-      );
+      this.db.predefinedTricks.bulkPut(tricks).then(() => {
+        console.log("added tricks to database from the csv");
+        this.db.versions.put({"key": "predefinedTricksVersion", "version": predefinedTricksVersion});
+      });
     });
   };
 
   // get single trick by id
-  getTrick = (id) => {
-    if (id < 1000) return this.db.userTricks.get(Number(id));
-    return Promise.all(
-      this.db.mainTricks.get(Number(id)),
-      this.db.mainAttributes.get(Number(id))
-    ).then((a) => {
-      if (a[1]) return Object({...a[0], ...a[1]});
-      return a[0];
-    });
-  };
+  getTrick = (id) => this.db.userTricks.get(Number(id)).then(userTrick => {
+    if (userTrick) return userTrick;
+    else return this.db.predefinedTricks.get(Number(id));
+  });
 
   // get list of all tricks
   getAllTricks = () => {
-    return Promise.all([
-      Promise.all([
-        this.db.mainTricks.toArray(),
-        this.db.mainAttributes.toArray()
-      ]).then((a) => {
-        const mainTricks = a[0];
-        const mainAttributes = a[1];
-        if (!mainAttributes) return mainTricks;
-
-        for (let i=0; i < mainAttributes.length; i++) {
-          const index = mainTricks.findIndex((trick) => trick.id === mainAttributes[i].id);
-          mainTricks[index] = Object({...mainTricks[index], ...mainAttributes[i]});
-        }
-        return mainTricks;
-      }),
-      this.db.userTricks.toArray()
-    ]).then((a) => a.flat());
+    return this.db.userTricks.toArray().then((userTricks) => {
+      const userKeys = userTricks.map(trick => trick.id);
+      // query all only predefinedTricks which don't have the same id as the userTricks
+      // and concat these to the userTricks
+      // also filter out tricks which are marked deleted
+      return this.db.predefinedTricks.where("id").noneOf(userKeys).toArray().then(preTricks => preTricks.concat(userTricks.filter(trick => !trick.deleted)));
+    });
   };
 
-  // create or update trick
-  saveTrick = (obj) => {
-    if (obj.id >= 1000) return this.db.mainTricks.put(obj);
-    return this.db.userTricks.put(obj);
-  };
+  // create or update userTrick
+  saveTrick = (trick) => this.db.userTricks.put(trick);
 
   // delete trick
-  deleteTrick = (id) => {
-    if (id < 1000) console.log("can't delete this trick");
-    return this.db.userTricks.delete(Number(id));
-  };
+  deleteTrick = (id) => this.db.userTricks.put({"id": Number(id), deleted: true});
 
```

Comment:
This means the user can not delete predefined tricks?

Reply:
no, the predefined lists are never modified. All the changes the user makes are noted in the user tricks.
If the user wants to delete a predefined trick, it's just marked as deleted in the user tables

Reply:
Ok, I see! Cool!